### ISSUE TITLE
[proposal] Extensibility

### ISSUE THREAD

**Issue by dandehavilland** [mozilla/metrics-graphics#344]
In an attempt to address #341, I propose extracting defaults and initializers from `data_graphic.js` and placing them into the corresponding chart files in `js/src/charts/*.js`. Each chart would then declare itself when loaded by making a call to a registration function:

``` JavaScript
// src/js/charts/missing.js
function() {
  // the chart definition which currently resides in `missing.js`
  var missing = function(args) {
    this.init = function() { ... },
    ...
  };

  // the chart-specific defaults extracted from `src/js/common/data_graphic.js`
  var defaults = {
    top: 40,
    bottom: 30,
    ...
  };

  // the chart-specific initializer, also extracted from `data_graphic.js`
  var initializer = function(args, commonDefaults) {
    // here we don't use `commonDefaults` (defaults.all), but it is available for other chart types
    args = merge_with_defaults(args, defaults);
    missing(args);
  };

  MG.register('missing-data', missing, initializer);
}.call();
```

`MG.register` deals with linking up this chart definition with the `data_graphic` function, which would then lookup the initializer based on `args.chart_type`.

The above example gives us two key improvements: 

1. All chart-specific code is kept together, and
2. It allows on-the-fly definition of new chart types, thus paving the way for extensibility.

I don't expect this to change the public-facing API so it could potentially be merged into a point release.

The next step would be to provide hooks and callbacks at key stages during the initialization and lifecycle of a chart to allow enhancements to the default behaviour via plugins, but I see that evolving as needs arise.

Any and all feedback welcome!

**Comment by almossawi** [comment]
Hi Dan, That sounds entirely reasonable. Thanks for putting this together.

@hamilton Sounds good to you?

**Comment by hamilton** [comment]
This is great. The API is very straightforward and will make life easier.

**Comment by hamilton** [comment]
I would agree that this is a pretty major refactoring. We can begin preparing some functions for said refactoring to make the transition seamless.

**Comment by dandehavilland** [comment]
I pushed ahead today and I've already made some progress on this - check my [addon-management](https://github.com/dandehavilland/metrics-graphics/tree/addon-management) - will be preparing a PR in the next day or 2. There are some additional considerations which I'll rase in another ticket tomorrow morning.

**Comment by hamilton** [comment]
Great work, dan. I've been tied up in other internal work but this is pretty exciting stuff.

**Comment by almossawi** [comment]
Targeting v2.5, seeing as we've merged #351.

**Comment by almossawi** [comment]
@dandehavilland Closing this issue, if that's ok with you :)
